### PR TITLE
feat: new protocol for chained functions, and added support for expli…

### PR DESCRIPTION
**Pull request by AppPear**
feat: new protocol for chained functions, and added support for explicit Y ranges. X coming as well

### CODE REVIEW

File: Sources/SwiftUICharts/Base/Grid/ChartGridShape.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        return path
+    }
+}
+
+struct ChartGridShape_Previews: PreviewProvider {
+    static var previews: some View {
+        Group {
+            ChartGridShape(numberOfHorizontalLines: 5, numberOfVerticalLines: 0)
+                .stroke()
+                .rotationEffect(.degrees(180), anchor: .center)
```
Comment (marekpivovarnik) [inline Sources/SwiftUICharts/Base/Grid/ChartGridShape.swift:20]: It also seems that these rotations are not necessary here. When I comment them the preview stays the same. But maybe I am just missing something important.

Comment (marekpivovarnik) [reply Sources/SwiftUICharts/Base/Grid/ChartGridShape.swift:20]: Ok, I think I got it. Are you trying to get `x` and `y` axis to be the same as you would intuitively expect? As in, point `(0.0)` is not top left corner but you want it to be lower left?

I think padding of line `.rotation3DEffect(.degrees(180), axis: (x: 0, y: 1, z: 0))` is not correct.

File: Sources/SwiftUICharts/Charts/LineChart/MarkerShape.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        return path
+    }
+}
+
+struct MarkerShape_Previews: PreviewProvider {
+    static var previews: some View {
+        Group {
+            MarkerShape(data: [(0, 0), (0.25, 0.5), (0.5,0.8), (0.75, 0.6), (1, 1)])
+                .stroke()
+                .rotationEffect(.degrees(180), anchor: .center)
```
Comment (marekpivovarnik) [inline Sources/SwiftUICharts/Charts/LineChart/MarkerShape.swift:16]: I think I would put it into `View` extension as it's keep being used on many places. Maybe something like this?

```
extension View {
    func toStandardCoordinateSystem() -> some View {
        self
            .rotationEffect(.degrees(180), anchor: .center)
            .rotation3DEffect(.degrees(180), axis: (x: 0, y: 1, z: 0))
    }
}
```

**Review by marekpivovarnik — APPROVED** [review]

File: Sources/SwiftUICharts/Base/Grid/ChartGridShape.swift (hunk 1 above)
Comment (AppPear) [reply Sources/SwiftUICharts/Base/Grid/ChartGridShape.swift:20]: done

File: Sources/SwiftUICharts/Charts/LineChart/MarkerShape.swift (hunk 2 above)
Comment (AppPear) [reply Sources/SwiftUICharts/Charts/LineChart/MarkerShape.swift:16]: done :)

File: Sources/SwiftUICharts/Base/Chart/ChartData.swift [hunk 3]
```diff
@@ -2,37 +2,78 @@ import SwiftUI
...
     }
 
-    var normalisedRange: Double {
-        (normalisedPoints.max() ?? 0.0) - (normalisedPoints.min() ?? 0.0)
+    var normalisedData: [(Double, Double)] {
+        Array(zip(normalisedValues, normalisedPoints))
+    }
+
+    var normalisedYRange: Double {
+        if let _ = rangeY {
```
[the remark this answers is not in the document]
Comment (AppPear) [reply Sources/SwiftUICharts/Base/Chart/ChartData.swift:44]: done :)

File: Sources/SwiftUICharts/Base/Chart/ChartData.swift [hunk 4]
```diff
@@ -2,37 +2,78 @@ import SwiftUI
...
+    public var rangeX: ClosedRange<Double>?
 
     var points: [Double] {
-        data.map { $0.1 }
+        data.filter { rangeX?.contains($0.0) ?? true }.map { $0.1 }
     }
 
-    var values: [String] {
-        data.map { $0.0 }
+    var values: [Double] {
```
[the remark this answers is not in the document]
Comment (AppPear) [reply Sources/SwiftUICharts/Base/Chart/ChartData.swift:13]: I would leave it points and values as not every chart has yValues (circle graphs so it wouldn't make sense there)

**Pull request merged**